fix: span url hostname not ip

astorm opened the pull request:
Fixes #2035

This PR add a `getUrlFromRequestAndOptions` function to the `lib/instrumentation/http-shared.js` module.  The `http-shared` module also uses this new function in place of the old `http-request-to-url` module to set the `url` property on a span.

The `getUrlFromRequestAndOptions` function accepts a `ClientRequest` object and a generic options object and uses the information from those objects to return the same final URL that the ClientRequest will use.  The options object is either 

1. The options object passed to the `request` method
   ```
    http.request(options)
   ```
       
2. Or, in the case of `http.request`'s second form    
   ```
    http.request(url, options)
   ```
   
   the `options` object [is merged with the URL information from `url`](https://github.com/elastic/apm-agent-nodejs/blob/842be710cbd8736bfff0867e2413ebb7cb0d229a/lib/instrumentation/http-shared.js#L146) to create a new object.
   
In addition to the usual test suite, this was tested across all supported node versions with [the following sludge-hammer of a bash script](https://gist.github.com/astorm/35ac16d0e2fb71b465f20381197e446f).

apmmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2039/detail/PR-2039/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2039/detail/PR-2039/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2039/detail/PR-2039/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2039/detail/PR-2039/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-agent-nodejs_2039.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #2039 opened`



* Start Time: 2021-04-10T00:12:42.216+0000


* Duration: 18 min 28 sec


* Commit: 6e5e8f33017ebcfcd41df9e49abeb4b2e65fd658



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 16614  |
| Skipped      | 0 |
| Total        | 16614   |


#### Trends :test_tube:


![Image of Build Times](https://apm-ci.elastic.co/job/apm-agent-nodejs/job/apm-agent-nodejs-mbp/job/PR-2039/buildTimeGraph/png)


![Image of Tests](https://apm-ci.elastic.co/job/apm-agent-nodejs/job/apm-agent-nodejs-mbp/job/PR-2039/1/testReport/history/countGraph/png?start=0&amp;end=25)


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->









### Steps errors [![29](https://img.shields.io/badge/29%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2039/detail/PR-2039/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


> Show only the first 10 steps failures

  
  
  
  
##### `Run Tests`
<ul>
<li>Took 5 min 51 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2039/runs/1/steps/546/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "10.0" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Run Tests`
<ul>
<li>Took 7 min 29 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2039/runs/1/steps/379/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "8" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Run Tests`
<ul>
<li>Took 5 min 52 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2039/runs/1/steps/558/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "8" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Run Tests`
<ul>
<li>Took 7 min 12 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2039/runs/1/steps/389/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "8" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Run Tests`
<ul>
<li>Took 5 min 43 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2039/runs/1/steps/550/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "8" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Run Tests`
<ul>
<li>Took 7 min 55 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2039/runs/1/steps/400/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "8.6" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Run Tests`
<ul>
<li>Took 5 min 37 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2039/runs/1/steps/530/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "8.6" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Run Tests`
<ul>
<li>Took 7 min 56 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2039/runs/1/steps/401/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "8.6" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Run Tests`
<ul>
<li>Took 5 min 38 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2039/runs/1/steps/542/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "8.6" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2039/runs/1/steps/749/log/?start=0">here</a></li>
<li>Description: <code>hudson.AbortException: script returned exit code 1</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-04-10T00:30:56.752Z]     Please visit http://docs.codecov.io/docs/supported-languages
[2021-04-10T00:30:57.197Z] Failed in branch Node.js-8
[2021-04-10T00:30:57.355Z] Recording test results
[2021-04-10T00:30:57.690Z] Digest: sha256:3423a5eea6a1151df079c5e94458dbd946cd53bcb75b47954a16af3137731ec7
[2021-04-10T00:30:57.690Z] Status: Downloaded newer image for node:12-alpine
[2021-04-10T00:30:58.402Z] [Checks API] No suitable checks publisher found.
[2021-04-10T00:30:58.460Z] Running in /var/lib/jenkins/workspace/ejs_apm-agent-nodejs-mbp_PR-2039/src/github.com/elastic/apm-agent-nodejs
[2021-04-10T00:30:58.510Z] [INFO] Codecov: Getting branch ref...
[2021-04-10T00:30:58.643Z] Masking supported pattern matches of $GITHUB_TOKEN
[2021-04-10T00:30:58.904Z] [INFO] Codecov: Sending data...
[2021-04-10T00:30:59.086Z] /tmp/.npm-global/bin/tap-xunit -> /tmp/.npm-global/lib/node_modules/tap-xunit/bin/tap-xunit
[2021-04-10T00:30:59.086Z] /tmp/.npm-global/bin/txunit -> /tmp/.npm-global/lib/node_modules/tap-xunit/bin/tap-xunit
[2021-04-10T00:30:59.086Z] + tap-xunit@2.4.1
[2021-04-10T00:30:59.086Z] added 21 packages from 21 contributors in 2.474s
[2021-04-10T00:30:59.506Z] + curl -sSLo codecov.sh https://codecov.io/bash
[2021-04-10T00:30:59.923Z] + bash codecov.sh
[2021-04-10T00:30:59.924Z] 
[2021-04-10T00:30:59.924Z]   _____          _
[2021-04-10T00:30:59.924Z]  / ____|        | |
[2021-04-10T00:30:59.924Z] | |     ___   __| | ___  ___ _____   __
[2021-04-10T00:30:59.924Z] | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
[2021-04-10T00:30:59.924Z] | |___| (_) | (_| |  __/ (_| (_) \ V /
[2021-04-10T00:30:59.924Z]  \_____\___/ \__,_|\___|\___\___/ \_/
[2021-04-10T00:30:59.924Z]                               Bash-20210309-2b87ace
[2021-04-10T00:30:59.924Z] 
[2021-04-10T00:30:59.924Z] 
[2021-04-10T00:30:59.924Z] ==> git version 2.17.1 found
[2021-04-10T00:30:59.924Z] ==> curl 7.58.0 (x86_64-pc-linux-gnu) libcurl/7.58.0 OpenSSL/1.1.1 zlib/1.2.11 libidn2/2.0.4 libpsl/0.19.1 (+libidn2/2.0.4) nghttp2/1.30.0 librtmp/2.3
[2021-04-10T00:30:59.924Z] Release-Date: 2018-01-24
[2021-04-10T00:30:59.924Z] Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp 
[2021-04-10T00:30:59.924Z] Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets HTTPS-proxy PSL 
[2021-04-10T00:30:59.924Z] ==> Jenkins CI detected.
[2021-04-10T00:30:59.924Z]     project root: .
[2021-04-10T00:30:59.924Z] --> token set from env
[2021-04-10T00:30:59.924Z]     Yaml found at: .codecov.yml
[2021-04-10T00:30:59.924Z] ==> Running gcov in . (disable via -X gcov)
[2021-04-10T00:31:00.069Z] Recording test results
[2021-04-10T00:31:00.186Z] ==> Python coveragepy not found
[2021-04-10T00:31:00.186Z] ==> Searching for coverage reports in:
[2021-04-10T00:31:00.186Z]     + .
[2021-04-10T00:31:00.186Z] --> No coverage report found.
[2021-04-10T00:31:00.186Z]     Please visit http://docs.codecov.io/docs/supported-languages
[2021-04-10T00:31:00.615Z] Failed in branch Node.js-10.0
[2021-04-10T00:31:01.098Z] [Checks API] No suitable checks publisher found.
[2021-04-10T00:31:01.169Z] Running in /var/lib/jenkins/workspace/ejs_apm-agent-nodejs-mbp_PR-2039/src/github.com/elastic/apm-agent-nodejs
[2021-04-10T00:31:01.226Z] [INFO] Codecov: Getting branch ref...
[2021-04-10T00:31:01.361Z] Masking supported pattern matches of $GITHUB_TOKEN
[2021-04-10T00:31:01.625Z] [INFO] Codecov: Sending data...
[2021-04-10T00:31:02.179Z] + curl -sSLo codecov.sh https://codecov.io/bash
[2021-04-10T00:31:02.575Z] + bash codecov.sh
[2021-04-10T00:31:02.575Z] 
[2021-04-10T00:31:02.575Z]   _____          _
[2021-04-10T00:31:02.575Z]  / ____|        | |
[2021-04-10T00:31:02.575Z] | |     ___   __| | ___  ___ _____   __
[2021-04-10T00:31:02.575Z] | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
[2021-04-10T00:31:02.575Z] | |___| (_) | (_| |  __/ (_| (_) \ V /
[2021-04-10T00:31:02.575Z]  \_____\___/ \__,_|\___|\___\___/ \_/
[2021-04-10T00:31:02.575Z]                               Bash-20210309-2b87ace
[2021-04-10T00:31:02.575Z] 
[2021-04-10T00:31:02.575Z] 
[2021-04-10T00:31:02.575Z] ==> git version 2.17.1 found
[2021-04-10T00:31:02.575Z] ==> curl 7.58.0 (x86_64-pc-linux-gnu) libcurl/7.58.0 OpenSSL/1.1.1 zlib/1.2.11 libidn2/2.0.4 libpsl/0.19.1 (+libidn2/2.0.4) nghttp2/1.30.0 librtmp/2.3
[2021-04-10T00:31:02.575Z] Release-Date: 2018-01-24
[2021-04-10T00:31:02.575Z] Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp 
[2021-04-10T00:31:02.575Z] Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets HTTPS-proxy PSL 
[2021-04-10T00:31:02.575Z] ==> Jenkins CI detected.
[2021-04-10T00:31:02.575Z]     project root: .
[2021-04-10T00:31:02.575Z] --> token set from env
[2021-04-10T00:31:02.575Z]     Yaml found at: .codecov.yml
[2021-04-10T00:31:02.576Z] ==> Running gcov in . (disable via -X gcov)
[2021-04-10T00:31:02.576Z] ==> Python coveragepy not found
[2021-04-10T00:31:02.576Z] ==> Searching for coverage reports in:
[2021-04-10T00:31:02.576Z]     + .
[2021-04-10T00:31:02.838Z] --> No coverage report found.
[2021-04-10T00:31:02.839Z]     Please visit http://docs.codecov.io/docs/supported-languages
[2021-04-10T00:31:03.284Z] Failed in branch Node.js-10
[2021-04-10T00:31:04.030Z] [INFO] getVaultSecret: Getting secrets
[2021-04-10T00:31:04.322Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-04-10T00:31:06.246Z] Stage "TAV Test" skipped due to earlier failure(s)
[2021-04-10T00:31:06.368Z] Stage "Edge Test" skipped due to earlier failure(s)
[2021-04-10T00:31:06.578Z] Stage "Nightly Test" skipped due to earlier failure(s)
[2021-04-10T00:31:06.603Z] Stage "Nightly Test - No async hooks" skipped due to earlier failure(s)
[2021-04-10T00:31:06.630Z] Stage "RC Test" skipped due to earlier failure(s)
[2021-04-10T00:31:06.658Z] Stage "RC Test - No async hooks" skipped due to earlier failure(s)
[2021-04-10T00:31:06.905Z] Failed in branch Nightly Test
[2021-04-10T00:31:06.934Z] Failed in branch Nightly Test - No async hooks
[2021-04-10T00:31:06.961Z] Failed in branch RC Test
[2021-04-10T00:31:06.992Z] Failed in branch RC Test - No async hooks
[2021-04-10T00:31:07.300Z] Stage "Integration Tests" skipped due to earlier failure(s)
[2021-04-10T00:31:07.421Z] Stage "Release" skipped due to earlier failure(s)
[2021-04-10T00:31:07.471Z] Stage "Release" skipped due to earlier failure(s)
[2021-04-10T00:31:07.671Z] Stage "Benchmarks" skipped due to earlier failure(s)
[2021-04-10T00:31:08.976Z] Running on worker-1225339 in /var/lib/jenkins/workspace/ejs_apm-agent-nodejs-mbp_PR-2039
[2021-04-10T00:31:09.072Z] [INFO] getVaultSecret: Getting secrets
[2021-04-10T00:31:09.333Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-04-10T00:31:11.761Z] + chmod 755 generate-build-data.sh
[2021-04-10T00:31:11.761Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2039/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2039/runs/1 FAILURE 1108132
[2021-04-10T00:31:11.761Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2039/runs/1/steps/?limit=10000 -o steps-info.json
[2021-04-10T00:31:12.460Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2039/runs/1/tests/?status=FAILED -o tests-errors.json
[2021-04-10T00:31:13.159Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2039/runs/1/log/ -o pipeline-log.txt

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

astorm commented:
Tests failing in `test/instrumentation/modules/http/outgoing.js` for nodes 12.0 and 14.0, but passing in v14.14 or v12.22.1.  Likely due to some subtly of localhost vs. 127.0.0.1.  Will investigate.

trentm commented on lib/instrumentation/http-shared.js:
```diff
@@ -260,3 +257,40 @@ function isAWSSigned (opts) {
+  req.agent = req.agent || {}
+
+  const port = options.port ? `:${options.port}` : ''
+  const host = req.host || options.hostname || options.host || 'localhost'
+  const protocol = req.protocol || req.agent.protocol
```
Would it be worth a comment (or overkill?) to note that req.host and req.protocol are only available in v14.5.0, v12.19.0 and later?

Also, `req.agent` isn't actually a documented attributes, but I agree this is the way to get to the protocol as a fallback. ... *or* we pass in the protocol given that we know if we are instrumenting `http.request` or `https.request`.

trentm replied:
BTW, forget the "... or we pass in the protocol" part. That's just me talking out loud. I'd only bother with this *if* we later go the route of trying to reconstruct the url from *just the given arguments*.

trentm replied:
Okay, I *do* have a baroque example of breaking this:
- if using a version of node before `req.protocol` was added
- and if passing in a custom `options.agent` that implements `function addRequest` (as required), but does *not* provide `agent.protocol`.

What happens currently is the `try { ... }` block below attempts to parse, e.g.: `undefined//localhost/get?foo=bar`, which throws and we return the sentinel (currently `false`).  I think that's reasonable for now. If we hear a complain, we can talk about passing in the protocol. 

Or we *could* pass in `moduleName` which we have already as the protocol.

astorm replied:
The comments are a good idea, I'll add them.

Also, since we already have the moduleName, passing it into `getUrlFromRequestAndOptions` and using it as a fallback value is a quick win.  I'll make these changes.

trentm commented on lib/instrumentation/http-shared.js:
```diff
@@ -260,3 +257,40 @@ function isAWSSigned (opts) {
+// @return string
+function getUrlFromRequestAndOptions (req, options) {
+  if (!req) {
+    return false
```
Returning false here results in `"url":false` in the span. E.g.:

```
{"span":{"name":"GET httpbin.org/Get","type":"external",..."duration":190.626,"context":{"http":{"method":"GET","status_code":404,"url":false}}...
```

astorm replied:
switched this up from false to undefined to ensure its dropped from the serialized span.

trentm commented on lib/instrumentation/http-shared.js:
```diff
@@ -260,3 +257,40 @@ function isAWSSigned (opts) {
+  try {
+    url = new URL(`${protocol}//${host}${port}${req.path}`)
+  } catch (e) {
+    // console.log(e)
```
Debug log to remove?

astorm replied:
✅ removing

trentm commented on lib/instrumentation/http-shared.js:
```diff
@@ -260,3 +257,40 @@ function isAWSSigned (opts) {
+    return false
+  }
+
+  return `${url.protocol}//${url.host}${url.pathname}${url.search}`
```
Why the step to `new URL`-parse the values we have, and then reconstruct it to a string again?

astorm replied:
Per discussion elsewhere, this is to remove the username:password portion of a hostname.  We rely on the URL object's parsing, and then reconstruct the URL to be just the parts we want.

astorm commented:
Per discussions, we've removed the "pass everything through a `URL` object to sanitize" in order to simplify the function (and adjusted the final return value to use local variables instead of properties from that URL object).  We've also added a  parameter to the function that lets us pass in the `moduleName` to use as a fallback protocol.  In tests we've moved the `username:password`  to the `auth` section of the options and added some test assertions that check for the presence of either string in the final URL. We've also added a test to ensure that port 80, if supplied, is represented in the URL.

trentm commented on lib/instrumentation/http-shared.js:
```diff
@@ -260,3 +256,47 @@ function isAWSSigned (opts) {
+  // req.host and req.protocol are node versions v14.5.0/v12.19.0 and later
+  const host = req.host || options.hostname || options.host || 'localhost'
+  const protocol = req.protocol || req.agent.protocol || fallbackProtocol
+  // let url
+  // try {
+  //   url = new URL(`${protocol}//${host}${port}${req.path}`)
+  // } catch (e) {
+  //   return undefined
+  // }
```
Drop these.

trentm commented on lib/instrumentation/http-shared.js:
```diff
@@ -191,13 +189,11 @@ exports.traceOutgoingRequest = function (agent, moduleName, method) {
-      httpRequestToUrl(req).then(_url => {
-        url = _url
-      }).catch(() => {
+      const url = getUrlFromRequestAndOptions(req, options, moduleName)
```
```suggestion
      const url = getUrlFromRequestAndOptions(req, options, moduleName + ':')
```

Need the colon, otherwise you get something like: `http//localhost/get` in the *very* special circumstances where `fallbackProtocol` is used. E.g. with node 10 or less, using this:

```
http.get(
  {
    hostname: 'localhost',
    path: '/get',
    // A custom agent that implements the minimum to pass muster, but does
    // *not* define `agent.protocol`.
    agent: {
      addRequest() {}
    }
  }
)
```

astorm replied:
got it -- I've made this change, and added the test I should have to test this particular case.
